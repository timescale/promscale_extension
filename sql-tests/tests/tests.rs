use insta::assert_snapshot;
use std::{
    env,
    mem::MaybeUninit,
    sync::{Mutex, Once},
};
use test_common::*;
use test_generator::test_resources;

const TESTDATA: &'static str = concat!(env!("CARGO_MANIFEST_DIR"), "/testdata");

fn init_lock() -> &'static Mutex<()> {
    static mut SINGLETON: MaybeUninit<Mutex<()>> = MaybeUninit::uninit();
    static ONCE: Once = Once::new();

    unsafe {
        ONCE.call_once(|| {
            let singleton = Mutex::new(());
            SINGLETON.write(singleton);
        });

        SINGLETON.assume_init_ref()
    }
}

// TODO fix how test_resources works in nexted workspaces
#[test_resources("sql-tests/testdata/*.sql")]
fn sql_tests(full_resource: &str) {
    let pg_blueprint = PostgresContainerBlueprint::new().with_testdata(TESTDATA);
    let test_pg_instance = new_test_instance_from_env(&pg_blueprint);
    let mut init_conn = test_pg_instance.connect();
    // This lock prevents multiple callers from running
    // CREATE EXTENSION simultaneously. In case of
    // promscale_extension concurrent CREATE EXTENSION causes
    // issues when migration incremental/003-users.sql executes.
    {
        let _init_lock = init_lock().lock().unwrap();
        init_conn
            .simple_query("CREATE EXTENSION promscale;")
            .expect("Unable to create extension promscale.");
    }

    let resource = if let Some((_, rest)) = full_resource.split_once('/') {
        rest
    } else {
        full_resource
    };
    let query_result = test_pg_instance.exec_sql_script(resource);
    assert_snapshot!(resource, query_result);
}
