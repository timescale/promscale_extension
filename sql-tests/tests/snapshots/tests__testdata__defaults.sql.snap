---
source: sql-tests/tests/tests.rs
expression: query_result
---
\set ON_ERROR_STOP 1
\echo make sure the view returns what we expect it to
make sure the view returns what we expect it to
select key, value
from _prom_catalog.initial_default dd
except
select key, value
from
(
    values
    ('chunk_interval'           , (INTERVAL '8 hours')::text),
    ('retention_period'         , (90 * INTERVAL '1 day')::text),
    ('metric_compression'       , (exists(select 1 from pg_catalog.pg_proc where proname = 'compress_chunk')::text)),
    ('trace_retention_period'   , (30 * INTERVAL '1 days')::text),
    ('ha_lease_timeout'         , '1m'),
    ('ha_lease_refresh'         , '10s'),
    ('epoch_duration'           , (INTERVAL '12 hours')::text),
    ('downsample'               , 'true'),
    ('downsample_old_data'      , 'false')
) x(key, value)
;
 key | value 
-----+-------
(0 rows)

\echo make sure the getter returns the expected value
make sure the getter returns the expected value
select _prom_catalog.get_default_value('chunk_interval') = (INTERVAL '8 hours')::text;
 ?column? 
----------
 t
(1 row)

select count(*) = 0 from _prom_catalog.default d where d.key = 'chunk_interval';
 ?column? 
----------
 t
(1 row)

\echo setting the default value to the same as the initial default
setting the default value to the same as the initial default
select _prom_catalog.set_default_value('chunk_interval', (INTERVAL '8 hours')::text);
 set_default_value 
-------------------
 
(1 row)

select count(*) = 1 from _prom_catalog.default d where d.key = 'chunk_interval';
 ?column? 
----------
 t
(1 row)

select _prom_catalog.get_default_value('chunk_interval') = (INTERVAL '8 hours')::text;
 ?column? 
----------
 t
(1 row)

\echo overriding the initial default
overriding the initial default
select _prom_catalog.set_default_value('chunk_interval', (INTERVAL '99 hours')::text);
 set_default_value 
-------------------
 
(1 row)

select count(*) = 1 from _prom_catalog.default d where d.key = 'chunk_interval';
 ?column? 
----------
 t
(1 row)

select _prom_catalog.get_default_value('chunk_interval') = (INTERVAL '99 hours')::text;
 ?column? 
----------
 t
(1 row)

\echo setting the default value BACK to the same as the initial default
setting the default value BACK to the same as the initial default
select _prom_catalog.set_default_value('chunk_interval', (INTERVAL '8 hours')::text);
 set_default_value 
-------------------
 
(1 row)

select count(*) = 1 from _prom_catalog.default d where d.key = 'chunk_interval';
 ?column? 
----------
 t
(1 row)

select _prom_catalog.get_default_value('chunk_interval') = (INTERVAL '8 hours')::text;
 ?column? 
----------
 t
(1 row)


